designs/NuGet-PackageFileService.md

```diff
@@ -0,0 +1,104 @@
+
+On the client side, when trying to fetch the icon, we call PackageFileService.GetPackageIconAsync, passing in the package identity.
+
+On the client side, when trying to fetch an embedded license, we call PackageFileService.GetEmbeddedLicenseAsync.
+
+On the service side, we then retrieve the appropriate file from the appropriate location based on the Uri stored in the MemoryCache and return the stream to the caller.
+
+#### IdentityToUri Memory Cache notes
+
+The settings of the IdentityToUri memory cache that the NuGetPackageFileService class hosts and maintains are identical to the metadata MemoryCache that the search service uses as part of it codespaces rebuilding:
+- 4 MB memory limit
+- 0 physical memory limit percentage
+- 2min polling interval
+
+Given those settings, we don't expect to frequently not find an entry that should be available in the memory cache...we have a Telemetry call to record how often this does happen.
+
+Details of how the memory cache gets populated is overed in the "paragraph above that starts with "On the service side".
+
+#### Issue: should GetEmbeddedLicenseAsync be more generic. (question from zkat)
+    zkat - I guess this is fine for now, but I'd really like it if now-or-eventually, this is abstracted away into a set of supported "Well Known Files"--for example, if we want to add Changelog or docs support in the future.
+
+    rrelyea - thought a bit about it... i'd imagine we could pass an enum into a method... packageIdentity, wellKnownEmbeddedFile
+
+    where wellKnownEmbeddedFiles might be readme, license and what else?
+    maybe icon...but that isn't always embedded, can also be a url outside of the package.
+    i lean towards not doing this in a generic way yet...as we don't know the full set of needs yet...and since the nuspec is on the server, we cannot just take a relative path, etc...
+
+    would love to discuss, hear other thoughts.
+
+### Other Changes
+
+Several classes had a PackageReader property that was of type Func<PackageReader>.
+Func<PackageReader> wasn't designed to be remoted nicely. In several places, I replaced the PackageReader property with a PackagePath property, to be used when appropriate for loading files from a package.
+
+## Test Strategy
+
+Added unit tests for NuGetPackageFileService and updated many existing tests for this new implementation of icons/licenses.
+
+## Future Work
+
+None critical.
+
+## Open Questions
+
+Any open questions not specifically called out in the design.
+
+## Considerations
+
+At one point, the design of this interface passed in URIs...we moved to PackageIdentity and storing in the MemoryCache to tighten security of the feature.
+
+We also, chose to only key off of PackageIdentity, instead of PackageIdentity and a package source...staying consistent with our treatment of unique package idenities as the same, regardless of where they come from.

```

Comment:
*  also, -> also
*  idenities -> identities

Reply:
fixed in next iteration.